dascore/proc/whitening.py

```diff
@@ -0,0 +1,161 @@
+
+    plot_spectrum(filt_patch.data[50,:],1/200,ax3)
+    ax3.set_title('Original data, distance = 50 m')
+    plot_spectrum(white_patch.data[50,:],1/200,ax4)
+    ax4.set_title('Whitened data, distance = 50 m')
+    plt.show()
+    """
+    patch_cop = patch.convert_units(distance="m").transpose("distance", "time")
+    time = patch_cop.coords.get_array("time")
+    dt = (time[1] - time[0]) / np.timedelta64(1, "s")
+
+    if freq_range is None:
+        freq_range = [0, 1 / 2 / dt]
+    else:
+        if not np.size(freq_range) == 2:
+            msg = "Frequency range must include two values"
+            raise ParameterError(msg)
+
+        if freq_range[0] < 0 or freq_range[1] < 0:
+            msg = "Minimal and maximal frequencies have to be non-negative"
+            raise ParameterError(msg)
+
+        if freq_range[0] >= freq_range[1]:
+            msg = "Frequency range must be increasing"
+            raise ParameterError(msg)
+
+    if freq_smooth_size is None:
+        freq_smooth_size = freq_range[1] - freq_range[0]
+    else:
+        if freq_smooth_size <= 0:
+            msg = "Frequency smoothing size must be positive"
+            raise ParameterError(msg)
+
+    # dt = (time[1] - time[0]) / np.timedelta64(1, "s")
+    (nchan, nt) = patch_cop.data.shape
+    nf = 2 ** (math.ceil(math.log(nt, 2)))
+    h_nf = int(nf / 2)
+    fft_d = np.zeros([nchan, nf], dtype=complex)
+    whitened_data = np.zeros([nchan, nt])
+
+    freqs = nft.fftfreq(nf, d=dt)
+    first_freq_ind = np.argmax((np.abs(freqs) - freq_range[0]) >= 0.0)
+    last_freq_ind = np.argmax((np.abs(freqs) - freq_range[1]) >= 0.0)
+    mean_window = math.floor(freq_smooth_size / (freqs[1] - freqs[0]))
+
+    if mean_window < 1:
+        msg = "Frequency smoothing size yields a smoothing window of size 0"
+        raise ParameterError(msg)
+    if last_freq_ind - first_freq_ind < 2:
+        msg = "Frequency range is too narrow"
+        raise ParameterError(msg)
+
+    if mean_window > last_freq_ind - first_freq_ind + 1:
+        msg = "Frequency smoothing size is larger than frequency range"
+        raise ParameterError(msg)
+
+    for i in range(nchan):
+        fft_d[i, :] = nft.fft(patch_cop.data[i, :], n=nf)
+
+    amp = np.abs(fft_d)
+    phase = np.angle(fft_d)
+
+    freqs[first_freq_ind:last_freq_ind]
+    conv_window = np.arange(mean_window) / mean_window
+
+    for i in range(nchan):
+        amp[i, first_freq_ind:last_freq_ind] = np.convolve(
+            amp[i, first_freq_ind:last_freq_ind], conv_window, mode="same"
+        )
+        sm_amp = amp[i, first_freq_ind:last_freq_ind]
+        amp[i, first_freq_ind:last_freq_ind] = np.divide(
+            sm_amp, np.abs(sm_amp), out=np.zeros_like(sm_amp), where=abs(sm_amp) != 0
+        )
+        amp[i, h_nf + 1 : nf] = np.flip(np.squeeze(amp[i, 1:h_nf]))
+        whitened_data[i, :] = np.real(nft.ifft(amp[i, :] * np.exp(1j * phase[i, :])))[
+            0:nt
+        ]
```

Comment:
Do we need to apply a tapering function to the edges (such as applying a cosine function)? I'm not sure how you handle boundary effects but see [whiten_2d](https://github.com/noisepy/NoisePy/blob/main/src/noisepy/seis/noise_module.py) function from NoisePy.

Reply:
Currently I only use the frequency range of interest for the calculation, and anything beyond it is lost. This is maybe why we see some noisier behavior at the edges. I am adding a taper param (in frequency units) so that the spectrum is computed using a wider range of frequencies (if possible), and the scaling value that we compute and divide by should be stabler.

If the users don't like the output spectrum near the edges, they can taper it themselves. I don't think this function should do it :)